pkg/controller/portworxdiag/portworxdiag.go

```diff
@@ -107,8 +153,351 @@ func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reco
+		logrus.WithError(err).Errorf("Failed to marshal json patch to JSON")
+		return err
+	}
+	err = c.client.Status().Patch(context.TODO(), diag, client.RawPatch(types.JSONPatchType, body))
+	if err != nil {
+		return fmt.Errorf("failed to add empty status object for PortworxDiag CR: %v", err)
+	}
+	return nil
+}
+
+func (c *Controller) patchClusterUUIDIfChanged(diag *diagv1.PortworxDiag, stc *corev1.StorageCluster) error {
+	if diag.Status.ClusterUUID == stc.Status.ClusterUID {
+		return nil
+	}
+
+	patch := []map[string]interface{}{
+		{
+			"op":    "add",
+			"path":  "/status/clusterUuid",
+			"value": stc.Status.ClusterUID,
+		},
+	}
+	body, err := json.Marshal(patch)
+	if err != nil {
+		logrus.WithError(err).Errorf("Failed to marshal json patch to JSON")
+		return err
+	}
+	err = c.client.Status().Patch(context.TODO(), diag, client.RawPatch(types.JSONPatchType, body))
+	if err != nil {
+		return fmt.Errorf("failed to update cluster UUID for PortworxDiag CR: %v", err)
+	}
+	return nil
+}
+
+func (c *Controller) addNodeStatusesIfMissing(diag *diagv1.PortworxDiag, nodeStatusesToAdd []*diagv1.NodeStatus) error {
+	if len(nodeStatusesToAdd) == 0 {
+		return nil
+	}
+
+	patch := []map[string]interface{}{}
+	if diag.Status.NodeStatuses == nil {
+		patch = append(patch, map[string]interface{}{
+			"op":    "add",
+			"path":  "/status/nodes",
+			"value": []diagv1.NodeStatus{},
+		})
+	}
+	for _, toAdd := range nodeStatusesToAdd {
+		patch = append(patch, map[string]interface{}{
+			"op":    "add",
+			"path":  "/status/nodes/-",
+			"value": toAdd,
+		})
+	}
+	body, err := json.Marshal(patch)
+	if err != nil {
+		logrus.WithError(err).Errorf("Failed to marshal json patch to JSON")
+		return err
+	}
+	err = c.client.Status().Patch(context.TODO(), diag, client.RawPatch(types.JSONPatchType, body))
+	if err != nil {
+		return fmt.Errorf("failed to update phase for PortworxDiag CR: %v", err)
+	}
+	return nil
+}
+
+func (c *Controller) patchPhaseIfChanged(diag *diagv1.PortworxDiag) error {
+	newPhase := getOverallPhase(diag)
+	if diag.Status.Phase == newPhase {
+		return nil
+	}
+
+	logrus.Infof("Changing PortworxDiag phase from %s to %s", diag.Status.Phase, newPhase)
+	patch := []map[string]interface{}{
+		{
+			"op":    "add",
+			"path":  "/status/phase",
+			"value": newPhase,
+		},
+	}
+	body, err := json.Marshal(patch)
+	if err != nil {
+		logrus.WithError(err).Errorf("Failed to marshal json patch to JSON")
+		return err
+	}
+	err = c.client.Status().Patch(context.TODO(), diag, client.RawPatch(types.JSONPatchType, body))
+	if err != nil {
+		return fmt.Errorf("failed to update phase for PortworxDiag CR: %v", err)
+	}
+	return nil
+}
+
+func (c *Controller) syncPortworxDiag(diag *diagv1.PortworxDiag) error {
+	logrus.Info("Enter syncPortworxDiag")
+	stc, err := c.fetchSTC()
+	if err != nil {
+		logrus.WithError(err).Error("Failed to find StorageCluster object")
+		return fmt.Errorf("failed to find StorageCluster object: %v", err)
+	}
+
```

Comment:
The Portworx Diag object doesn't have an overall status field (though it may be good to add one). Should we add it, or just set it to failed and maybe add an event instead?

Reply:
It would be good to have a message field in `PortworxDiagStatus` and populate that. Let's add that field, mark the phase as Failed and also raise the event.